File: algorithms/DLS.py
# Depth Limit search
from collections import deque
import networkx as nx
from typing import Any, Optional

def dls(G: nx.Graph, source, target, d = 50) -> Optional[tuple[list[Any], float, list[Any]]]:
    if source not in G or target not in G:
        return None
    
    q = deque([source])
    parent = {source: None}
    depth = {source: 0}
    visit_order: list[Any] = []

    while q:
        u = q.pop()
        visit_order.append(u)

        if u == target:
            path = []
            cur = u
            while cur is not None:
                path.append(cur)
                cur = parent[cur]
            path.reverse()

            total_weight = 0.0
            if len(path)>=2 :
                for i in range(len(path)- 1):
                    a, b = path[i], path[i+1]
                    if G.has_edge(a,b):
                        data = G.get_edge_data(a,b) or {}
                        if isinstance(data, dict) and any(isinstance(v, dict) for v in data.values()):
                            first_key = next(iter(data))
                            attrs = data[first_key] or {}
                        else:
                            attrs = data
                        w = attrs.get('weight', 1.0)
                    else:
                        w = 1.0
                    try:
                        total_weight += float(w)
                    except Exception:
                        total_weight += 1.0
            return path, total_weight, visit_order
        
        for v in reversed(list(G.neighbors(u))):
            if v not in parent and depth[u] < d:
                depth[v] = depth[u]+1
                parent[v] = u
                q.append(v)

    return None



    

File: webapp/src/algorithms/dls.js
// Depth Limited Search algorithm with step-by-step visualization

import { getNeighbors } from '../utils/graphUtils';

/**
 * DLS algorithm implementation with event emission for visualization
 * @param {Object} graph - Graph object with nodes and edges
 * @param {string} source - Source node id
 * @param {string} target - Target node id
 * @param {boolean} isDirected - Whether the graph is directed
 * @param {Object} options - Options object (can contain depthLimit)
 * @returns {Object} - Result with path, cost, steps, and visited order
 */
export function dls(graph, source, target, isDirected = false, options = {}) {
  const depthLimit = options.depthLimit || 50;
  const steps = [];
  const visited = new Set();
  const parent = new Map();
  const depth = new Map();
  const stack = [source];
  const visitOrder = [];

  parent.set(source, null);
  depth.set(source, 0);

  // Initial step
  steps.push({
    type: 'init',
    current: null,
    queue: [...stack],
    visited: [...visited],
    frontier: [...stack],
    parent: Object.fromEntries(parent),
    depthLimit,
    message: `Starting DLS from node ${source} with depth limit ${depthLimit}`,
  });

  while (stack.length > 0) {
    const current = stack.pop();

    if (visited.has(current)) continue;
    visited.add(current);
    visitOrder.push(current);

    const currentDepth = depth.get(current);

    // Expand step
    steps.push({
      type: 'expand',
      current,
      queue: [...stack],
      visited: [...visited],
      frontier: [...stack],
      parent: Object.fromEntries(parent),
      currentDepth,
      message: `Expanding node ${current} at depth ${currentDepth}`,
    });

    // Check if we reached the target
    if (current === target) {
      const path = reconstructPath(parent, target);
      const cost = calculatePathCost(graph, path);

      steps.push({
        type: 'goal',
        current,
        path,
        cost,
        visited: [...visited],
        message: `Found target ${target}! Path length: ${path.length}, Cost: ${cost}`,
      });

      return {
        found: true,
        path,
        cost,
        steps,
        visitOrder,
        nodesVisited: visited.size,
      };
    }

    // Explore neighbors if within depth limit
    if (currentDepth < depthLimit) {
      const neighbors = getNeighbors(graph, current, isDirected);
      // Reverse to maintain left-to-right order when popping from stack
      for (const { node: neighbor } of [...neighbors].reverse()) {
        if (!visited.has(neighbor) && !parent.has(neighbor)) {
          parent.set(neighbor, current);
          depth.set(neighbor, currentDepth + 1);
          stack.push(neighbor);

          steps.push({
            type: 'frontier',
            current,
            neighbor,
            queue: [...stack],
            visited: [...visited],
            frontier: [...stack],
            parent: Object.fromEntries(parent),
            neighborDepth: currentDepth + 1,
            message: `Adding ${neighbor} to stack at depth ${currentDepth + 1}`,
          });
        }
      }
    } else {
      steps.push({
        type: 'depth_limit',
        current,
        visited: [...visited],
        message: `Depth limit ${depthLimit} reached at node ${current}`,
      });
    }
  }

  // No path found
  steps.push({
    type: 'no_path',
    visited: [...visited],
    message: `No path found from ${source} to ${target} within depth limit ${depthLimit}`,
  });

  return {
    found: false,
    path: [],
    cost: Infinity,
    steps,
    visitOrder,
    nodesVisited: visited.size,
  };
}

/**
 * Reconstruct path from parent map
 */
function reconstructPath(parent, target) {
  const path = [];
  let current = target;
  while (current !== null) {
    path.unshift(current);
    current = parent.get(current);
  }
  return path;
}

/**
 * Calculate total cost of a path
 */
function calculatePathCost(graph, path) {
  let cost = 0;
  for (let i = 0; i < path.length - 1; i++) {
    const edge = graph.edges.find(
      e => (e.from === path[i] && e.to === path[i + 1]) ||
           (e.from === path[i + 1] && e.to === path[i])
    );
    cost += edge ? edge.weight : 1;
  }
  return cost;
}

export default dls;
